File: solutions/prolog/knapsack/knapsack.pl
maximum_value(Items, Capacity, Value) :-
    maximum_value(Items, [], Capacity, Value).

weight(item(Weight, _), Weight).
weights([], 0).
weights([X|Xs], Weight) :-
  weight(X, W),
  weights(Xs, Ws),
  Weight is W + Ws.

value(item(_, Value), Value).
values([], 0).
values([X|Xs], Value) :-
  value(X, V),
  values(Xs, Vs),
  Value is V + Vs.

maximum_value([], [], _, 0).
maximum_value(_, Cur, Capacity, Value) :-
  weights(Cur, W),
  W > Capacity,
  Value is 0.
maximum_value([], Cur, Capacity, Value) :-
  weights(Cur, W),
  W =< Capacity,
  values(Cur, Value).
maximum_value([X|Xs], Cur, Capacity, Value) :-
  weights(Cur, W),
  W =< Capacity,
  values(Cur, V1),
  maximum_value(Xs, [X|Cur], Capacity, V2),
  maximum_value(Xs, Cur, Capacity, V3),
  Value is max(V1, max(V2, V3)).

File: solutions/java/knapsack/src/main/java/Knapsack.java
import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

class Knapsack {

    int maximumValue(int maximumWeight, List<Item> items) {
        return maximumValue(maximumWeight, new HashSet<>(), items);
    }

    int maximumValue(int maximumWeight, Set<Item> current, List<Item> rem) {
        int weight = current.stream().mapToInt(Item::getWeight).sum();
        if (weight > maximumWeight)
            return 0;

        var currentValue = current.stream().mapToInt(Item::getValue).sum();

        if (rem.isEmpty()) // Nothing more to see
            return currentValue;
        var withSelf = new HashSet<>(current);
        withSelf.add(rem.get(0));
        var tail = tail(rem);

        return Math.max(
            maximumValue(maximumWeight, current, tail),
            Math.max(
                maximumValue(maximumWeight, withSelf, tail),
                currentValue
            )
        );
    }

    // Java is dogshit
    private static List<Item> tail(List<Item> list) {
        return list.subList(1, list.size());
    }
}